pybop/costs/_likelihoods.py

```diff
@@ -319,26 +240,48 @@ def compute(
         calculate_grad: bool = False,
     ) -> Union[float, tuple[float, np.ndarray]]:
         """
-        Calculate the posterior cost for a given set of parameters.
+        Calculate the posterior cost for a given forward model prediction.
 
         Parameters
         ----------
-        x : array-like
-            The parameters for which to evaluate the cost.
-        grad : array-like, optional
-            An array to store the gradient of the cost function with respect
-            to the parameters.
+        y : dict
+            The data for which to evaluate the cost.
+        dy : np.ndarray, optional
+            The correspond sensitivities in the data.
+        calculate_grad : bool, optional
+            Whether to calculate the gradient of the cost function.
 
         Returns
         -------
-        float
-            The posterior cost.
+        Union[float, Tuple[float, np.ndarray]]
+            The posterior cost, and optionally the gradient.
         """
         # Verify we have dy if calculate_grad is True
         self.verify_args(dy, calculate_grad)
 
         if calculate_grad:
-            log_prior, dp = self._prior.logpdfS1(self.parameters.current_value())
+            if hasattr(self._prior, "logpdfS1"):
+                log_prior, dp = self._prior.logpdfS1(self.parameters.current_value())
+            else:
+                # Compute log prior first
+                log_prior = self._prior.logpdf(self.parameters.current_value())
+
+                # Compute a finite difference approximation of the gradient of the log prior
+                delta = self.parameters.initial_value() * self.gradient_step
+                dp = []
+
+                for parameter, step_size in zip(self.parameters, delta):
+                    param_value = parameter.value
+                    upper_value = param_value * (1 + step_size)
+                    lower_value = param_value * (1 - step_size)
+
+                    log_prior_upper = parameter.prior.logpdf(upper_value)
+                    log_prior_lower = parameter.prior.logpdf(lower_value)
+
+                    gradient = (log_prior_upper - log_prior_lower) / (
```

Comment:
That's understandable, thanks for the suggestion. I'm conscious that this is a fallback method, and not sure how much functionality is worthwhile here. Either way, I think this is probably worth the time. I'll open an issue to investigate this further.

As an aside, the addition of #481 opens up a way to auto-diff the priors, so this could be removed.